### PR TITLE
feat(polar): expand Polar AccessLink integration with cardio load, intraday samples, and temperature

### PR DESCRIPTION
**Pull request by CodeWithCJ**
> [!TIP]
> **Help us review and merge your PR faster!**
> Please ensure you have completed the **Checklist** below.
> For **Frontend** changes, please run `pnpm run validate` to check for any errors.
> PRs that include tests and clear screenshots are highly preferred!
> Note: AI-generated descriptions must be manually edited for conciseness. Do not paste raw AI summaries.

## Description

**What problem does this PR solve?**
Expands the Polar AccessLink integration to ingest and visualize Cardio Load (strain, tolerance, and ACWR ratio), Intraday Samples (5-minute overnight HRV series, Continuous Heart Rate, SpO2 biosensing spot tests), and Temperature (Body Temperature vitals and Skin Temperature series).

**How did you implement the solution?**
- Added Polar AccessLink v3 API fetchers for Cardio Load, Continuous HR, SpO2, Body Temperature, and Skin Temperature in `polarService.ts`.
- Implemented data processors in `polarDataProcessor.ts` to compute metrics, map sleep-average SpO2, and store series data into `health_metric_samples`, `daily_health_metrics`, and `vitals_entries`.
- Added database migration `20260918180000_widen_health_metric_samples_for_skin_temperature.sql` and updated shared Zod schemas (`HealthMetricSamples.zod.ts`).
- Added a Training Load section to `DailyHealthMetricsCard.tsx`, added `IntradaySamplesChart.tsx`, and updated `HRVCard.tsx` and `SpO2Card.tsx` with toggles for overnight 5-minute series and spot test views.
- Added comprehensive unit tests in `polarDataProcessor.test.ts` and frontend card tests.

Linked Issue: Closes #

## How to Test

1. Check out this branch and run `pnpm start` in `SparkyFitnessServer` and `pnpm dev` in `SparkyFitnessFrontend`.
2. Run backend tests: `cd SparkyFitnessServer && pnpm exec vitest run tests/polarDataProcessor.test.ts`
3. Run frontend tests: `cd SparkyFitnessFrontend && pnpm test src/tests/components/DailyHealthMetricsCard.test.tsx`
4. Trigger a Polar sync or replay sync and verify the Training Load tile appears on the Diary page, and the Night Series / Spot Tests toggles appear on the HRV and SpO2 report cards.

## PR Type

- [ ] Issue (bug fix)
- [x] New Feature
- [ ] Refactor
- [ ] Documentation

## Checklist

**All PRs:**

- [x] **[MANDATORY - ALL] Integrity & License**: I certify this is my own work, free of malicious code, and I agree to the [License terms](https://github.com/CodeWithCJ/SparkyFitness/blob/main/LICENSE).

**New features only:**

- [x] **[MANDATORY for new feature] Alignment**: I have raised a GitHub issue and it was reviewed/approved by maintainers or it was approved on Discord.

**Frontend changes (`SparkyFitnessFrontend/`):**

- [x] **[MANDATORY for Frontend changes] Quality**: I have run `pnpm run validate` and it passes.
- [x] **[MANDATORY for Frontend changes] Translations**: I have only updated the English (`en`) translation file.

**Backend changes (`SparkyFitnessServer/`):**

- [x] **[MANDATORY for Backend changes] Code Quality**: I have run typecheck, lint, and tests. New files use TypeScript, new endpoints have Zod schemas, and new endpoints include tests.
- [x] **[MANDATORY for Backend changes] Database Security**: I have updated `rls_policies.sql` for any new user-specific tables.

**UI changes (components, screens, pages):**

- [ ] **[MANDATORY for UI changes] Screenshots**: I have attached Before/After screenshots below.

**Mobile changes (`SparkyFitnessMobile/`):**

- [ ] **[MANDATORY for Mobile changes] Tested on device or emulator**: I have verified the changes work on iOS or Android.

## Screenshots

<details>
<summary>Click to expand</summary>

### Before

N/A

### After

N/A

</details>

## Notes for Reviewers

- Database migration `20260918180000_widen_health_metric_samples_for_skin_temperature.sql` widens the existing check constraint on `health_metric_samples` to support `skin_temperature`.
- Intraday sample curves use generic `health_metric_samples` queries, making the visualization component reusable for other wearables.

## Summary by CodeRabbit

* **New Features**
  * Added Training Load metrics, including acute, chronic, weekly load, and ACWR.
  * Expanded Polar wearable support for cardio load, continuous heart rate, SpO2, body temperature, skin temperature, and intraday HRV.
  * Added overnight HRV series and SpO2 spot-test views with interactive chart toggles.
  * Added intraday metric charts with latest, average, minimum, and maximum statistics.
  * Charts now respect the configured timezone and time format.
* **Localization**
  * Added labels for training load, skin temperature, table statistics, HRV trends, and sleep-health views.

### CODE REVIEW

File: SparkyFitnessServer/integrations/polar/polarDataProcessor.ts [hunk 1]
```diff
@@ -790,8 +800,597 @@ async function processPolarNightlyRecharge(
...
+      }
+
+      if (hrvSamples.length > 0) {
+        await upsertSamplesByDay(
+          userId,
+          createdByUserId,
+          'hrv',
+          POLAR_HEALTH_PROVIDER,
+          hrvSamples
+        );
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/integrations/polar/polarDataProcessor.ts:946]: Confirmed and fixed in dac4aaeca — this was the worst of the batch, and it was a bug the midnight-splitting fix introduced.

Because a night straddles midnight, each recharge record is a partial-day write for *both* days it touches, and consecutive nights share a bucket: night N's morning samples and night N+1's evening samples both land on day N. With the default `replace` mode the later night wiped the earlier one.

Measured against the reference payload, processing all seven nights:

```
night 1: {"2026-09-08":12,"2026-09-09":81}
night 2: {"2026-09-09":22,"2026-09-10":83}
...
day buckets written by more than one call:
  2026-09-09, 09-10, 09-11, 09-12, 09-13, 09-14
```

Six of seven days were being written twice, so night N's ~80 morning readings were replaced by night N+1's ~20 evening ones. Now written with `mode: 'merge'` and the night's own `{startMs, endMs}` window, which is exactly the case the writer's own docs describe. Covered by a test asserting both the merge mode and the shared-day overlap.

File: SparkyFitnessServer/integrations/polar/polarDataProcessor.ts [hunk 2]
```diff
@@ -790,8 +800,597 @@ async function processPolarNightlyRecharge(
...
+    } else if (testTime) {
+      timestamp = new Date(parsePolarToUTC(testTime) as string);
+    } else {
+      continue;
+    }
+
+    if (!Number.isFinite(timestamp.getTime())) continue;
+
+    const entryDate =
+      getVal(item, 'date') || timestamp.toISOString().split('T')[0];
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/integrations/polar/polarDataProcessor.ts]: Confirmed and fixed in dac4aaeca. Two sites, both wrong in the same way:

- SpO2 derived `entry_date` with `timestamp.toISOString().split('T')[0]`, the exact pattern AGENTS.md calls out. It now prefers the per-test `time_zone_offset` the spec supplies — better than the account timezone, since it reflects where the device actually was — and falls back to the user's zone.
- Body temperature used `instantToDay(..., 'UTC')`, which I added; it now uses the user's timezone.

A test covers the boundary case: a reading at 01:00 in UTC+2 is 23:00 UTC the previous day and must still bucket on the local date.

File: SparkyFitnessFrontend/src/utils/dailyHealthMetrics.ts [hunk 3]
```diff
@@ -10,6 +10,8 @@ const DISPLAYABLE_FIELDS = [
   'resting_heart_rate',
   'vo2_max',
   'training_readiness_score',
+  'acute_training_load',
+  'weekly_training_load',
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessFrontend/src/utils/dailyHealthMetrics.ts:14]: Confirmed and fixed in dac4aaeca, together with the related finding on DailyHealthMetricsCard.tsx:96 — same root cause.

The Training Load tile renders `acwr_ratio` and `chronic_training_load`, but `DISPLAYABLE_FIELDS` listed neither. `processPolarCardioLoad` writes the row when *any* of strain/tolerance/ratio is present, so Polar can produce a row with tolerance and ratio but no strain; the selector would have skipped it and hidden the card entirely. Both fields added.

File: SparkyFitnessFrontend/src/components/Health/DailyHealthMetricsCard.tsx [hunk 4]
```diff
@@ -84,19 +85,26 @@ export const DailyHealthMetricsCard: React.FC<DailyHealthMetricsCardProps> = ({
...
-  // render the rest as a wall of "--". These are the same five fields the card's
+  // render the rest as a wall of "--". These are the same fields the card's
   // own visibility keys on, so the card shows exactly when a tile would.
   const showBodyBattery = metrics.body_battery_highest != null;
   const showStress = metrics.avg_stress_level != null;
   const showRestingHr = metrics.resting_heart_rate != null;
   const showVo2Max = metrics.vo2_max != null;
   const showReadiness = metrics.training_readiness_score != null;
+  const showTrainingLoad =
+    metrics.acute_training_load != null || metrics.weekly_training_load != null;
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessFrontend/src/components/Health/DailyHealthMetricsCard.tsx:96]: Same root cause as the finding on dailyHealthMetrics.ts:14 — fixed there in dac4aaeca by adding `chronic_training_load` and `acwr_ratio` to `DISPLAYABLE_FIELDS`, so a Polar row carrying tolerance/ratio without strain still renders the tile.

File: SparkyFitnessFrontend/src/components/Health/IntradaySamplesChart.tsx [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+          const d = new Date(isoTime);
+          if (!Number.isFinite(d.getTime())) continue;
+
+          // Format time as HH:mm or hh:mm a
+          const is12Hour = timeFormat === '12h';
+          const hours = d.getHours();
+          const minutes = d.getMinutes().toString().padStart(2, '0');
+          const displayTime = is12Hour
+            ? `${hours % 12 || 12}:${minutes} ${hours >= 12 ? 'PM' : 'AM'}`
+            : `${hours.toString().padStart(2, '0')}:${minutes}`;
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessFrontend/src/components/Health/IntradaySamplesChart.tsx]: Confirmed and fixed in dac4aaeca. Verified the premise first: `formatTime` is exported from PreferencesContext (line 212) and `StressChart.tsx:37` uses it exactly as described.

Both new charts hand-rolled the formatting with `getHours()`/`getMinutes()`, which reads the browser timezone rather than the configured one — wrong for anyone whose account timezone differs from their machine, and a duplicate of logic that already exists. `IntradaySamplesChart` and `HRVCard` now call `formatTime`, with the memo dependencies updated.

File: SparkyFitnessServer/db/migrations/20260918180000_widen_health_metric_samples_for_skin_temperature.sql [hunk 6]
```diff
@@ -0,0 +1,8 @@
+-- Widen chk_health_metric_samples_metric to include 'skin_temperature'
+ALTER TABLE health_metric_samples
+    DROP CONSTRAINT IF EXISTS chk_health_metric_samples_metric;
+
+ALTER TABLE health_metric_samples
+    ADD CONSTRAINT chk_health_metric_samples_metric CHECK (metric IN (
+        'heart_rate', 'hrv', 'respiration', 'spo2', 'stress', 'body_battery', 'skin_temperature'
+    ));
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/db/migrations/20260918180000_widen_health_metric_samples_for_skin_temperature.sql:8]: Not actioning this one — the PR already does what the second half of the comment asks.

`db_schema_backup.sql` is deliberately untouched here. AGENTS.md is explicit: "Never manually edit `db_schema_backup.sql` — after merge, CI regenerates it from the migrations and opens an automated sync PR (`.github/workflows/schema-backup.yml`)." So the backup still showing six metrics is the expected state for an open PR; it updates after merge, not in it.

The headline asks to regenerate the checked-in backup while the body says to let the workflow do it and not edit it manually. Those point in opposite directions, and the repo convention settles it.

[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/db/migrations/20260918180000_widen_health_metric_samples_for_skin_temperature.sql:8]: Accepting the lock rather than restructuring, and the reasoning is worth recording.

The mechanics are right: `ADD CONSTRAINT` takes ACCESS EXCLUSIVE and scans existing rows, and because the runner sends the file through a single `client.query(sql)` the usual `NOT VALID` + `VALIDATE` split would not release the lock between statements. Splitting across two migration files would be the only way to get that benefit.

It is not worth it here:

- **No traffic is being served during the scan.** Migrations run from `index.ts` before any application module is imported — `initializeDatabase()` completes before `SparkyFitnessServer.ts` (and therefore the HTTP listener) is loaded. There are no writes to block on the instance doing the migrating, and concurrent startups are already serialised by the advisory lock.
- **Every existing row already satisfies the new constraint.** This widens the allowed set from six values to seven, so the scan is a pass over rows that cannot fail.
- `health_metric_samples` is jsonb-bucketed at one row per user/metric/day/provider, so it is far smaller than a per-sample table would be.

Flagging it as a real consideration for a future constraint that *narrows* a set, or for a table with per-sample rows — neither applies to this migration.
